tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Session.java

```diff
@@ -485,6 +475,26 @@ public void run(Op op) {
      * this field may be replaced by more type-safe equivalents at any time.
      */
     public RunMetadata metadata;
+
+    Run(List<Tensor<?>> outputs, RunMetadata metadata) {
+      this.outputs = outputs;
+      this.metadata = metadata;
+    }
+
+    Run(List<Tensor<?>> outputs) {
+      this.outputs = outputs;
+    }
+
+    public Tensor<?> get(int index) {
+      return this.outputs.get(index);
+    }
+
+    @Override
+    public void close() {
+      for (Tensor<?> tensor : this.outputs) {
+        tensor.close();
```

Comment:
Just noticed this, if we auto-close tensors resulting from a run, we should allow the user to retain a reference to one or more of them so the tensor will survive beyond the life of `Run` if needed.

This could be done by adding a utility method like `Tensor.retain()` which will call `tensorHandle.retainReference()` internally... or we remove this change from this actual PR and do it later, your choice.

Reply:
Great point! I think it makes sense to handle this outside this PR, as this functionality isn't used here. I reverted `Session.java` to the (new) master.